ENHANCE: Use anupq for p-group isom.

hulpke opened the pull request:
If the package is available

## Text for release notes

If the anupq package is installed it will be used for isomorphism of p groups.

fingolfin commented on lib/morpheus.gi:
```diff
@@ -2859,6 +2897,10 @@ local m;
     if m<>fail then return m;fi;
   fi;
 
+  if Size(G)>50 and IsPGroup(G) and LoadPackage("anupq")=true then
```
So far the policy for the GAP library always has been that we don't load packages, we only use them if already loaded. That would suggest to do this:
```suggestion
  if Size(G)>50 and IsPGroup(G) and IsPackageMarkedForLoading("anupq","")=truethen
```

Of course we can revisit this policy and discuss if this is still what we want, but perhaps that should be done on the GAP mailing list and not in a pull request discussion?

fingolfin replied:
Note that we do load `anupq` by default if it is compiled, so in practice there won't be much of a difference either way...

fingolfin replied:
Oh wait, that's actually wrong, I confused it with `autpgrp`. So `anupq` is *not* loaded by default. Apologies for the confusion.

fingolfin commented on lib/morpheus.gi:
```diff
@@ -2818,6 +2818,44 @@ local d,iso,a,b,c,o,s,two,rt,r,z,e,y,re,m,gens,cnt,lim,p,
+  imgs:=List(MappingGeneratorsImages(eG)[2],
+    x->MappedWord(UnderlyingElement(x),
+      FreeGeneratorsOfFpGroup(fG),GeneratorsOfGroup(pc)));
+  eG:=GroupHomomorphismByImages(G,pc,MappingGeneratorsImages(eG)[1],imgs);
```
Maybe use the `NC` version here and below? Or do you want the extra checks?
```suggestion
  eG:=GroupHomomorphismByImagesNC(G,pc,MappingGeneratorsImages(eG)[1],imgs);
```

hulpke replied:
Since an error (maybe in how I interpret the PQ results) would be disastrous, and the overall calculation is costly, the test here did seem to be worth doing, thus deliberately no NC.

hulpke replied:
Let me just add, that indeed this already caught a problem (which is fixed in a second commit)

fingolfin commented on lib/morpheus.gi:
```diff
@@ -2818,6 +2818,44 @@ local d,iso,a,b,c,o,s,two,rt,r,z,e,y,re,m,gens,cnt,lim,p,
+  imgs:=List(MappingGeneratorsImages(eH)[2],
+    x->MappedWord(UnderlyingElement(x),
+      FreeGeneratorsOfFpGroup(fH),GeneratorsOfGroup(pc)));
+  eH:=GroupHomomorphismByImages(H,pc,MappingGeneratorsImages(eH)[1],imgs);
```
```suggestion
  eH:=GroupHomomorphismByImagesNC(H,pc,MappingGeneratorsImages(eH)[1],imgs);
```

fingolfin commented on lib/morpheus.gi:
```diff
@@ -2818,6 +2818,44 @@ local d,iso,a,b,c,o,s,two,rt,r,z,e,y,re,m,gens,cnt,lim,p,
+  p:=p[1][1];
+  eG:=CallFuncList(ValueGlobal("EpimorphismPqStandardPresentation"),[G]:
+    Prime:=p);
+  eH:=CallFuncList(ValueGlobal("EpimorphismPqStandardPresentation"),[H]:
```
Seems this function is not actually tested by the test suite, at least in CI -- there very well may be an existing test triggering this function, but of course only if `anupq` can be loaded... but I think we don't actually compile `anupq` when running the CI tests. I guess we could change that by adding `anupq` to line 44 in `dev/ci-download-pkgs.sh`

hulpke replied:
Even with ANUPQ this can be quickly quite costly and thus did not seem a good idea to add. I'm happy to provide a test if it is requested.

fingolfin reviewed (commented):
Thank you! Added some minor remarks, but the `LoadPackage` bit seems bigger?

hulpke commented:
> Thank you! Added some minor remarks, but the `LoadPackage` bit seems bigger?

OK, I've changed this as suggested to stay within the existing policy. This is, however, a situation where a calculation goes from impossible to doable in 5 minutes, and the package need is not clear to the naive user. Should one issue a warning?
